### ISSUE TITLE
Providing X509_LOOKUP method 'from memory' in addition to 'from file'

### ISSUE THREAD

**Issue by trueqbit** [openssl/openssl#13134]
In order to verify a CA bundle file and/or certificates I am looking for a way to load the necessary certificate data from memory instead of from a file.

I am aware that it's possible to use `X509_STORE_add_cert()`/`X509_STORE_add_crl()` directly or add a custom `X509_LOOKUP_METHOD` method.
  I hope, however, that the requirement to build a certificate store from memory rather than the filesystem, and the low but apparent complexity, justify extending the openssl library itself.
  For reference, [Mr Ed on SO shared](https://stackoverflow.com/a/17013082/279251) this approach to someone already years ago.

Using it would look something like:

    X509_STORE *setup_verify(const char *CAdata)
    {
        X509_STORE *store = X509_STORE_new();

        X509_LOOKUP *lookup = X509_STORE_add_lookup(store, X509_LOOKUP_mem());

        X509_LOOKUP_load_mem(lookup, CAdata, X509_FILETYPE_PEM);

        ERR_clear_error();
        return store;
     end:
        X509_STORE_free(store);
        return NULL;
    }

Loading from memory has a lot in common with loading from a file as implemented in _openssl/crypto/x509/by_file.c_.

The one thing that both methods share can be factored out into __functions dealing with a BIO structure, which seems worth in its own respect anyway__.
  

Current file-based code could be duplicated into its own _by_mem.c_ or existing code in `by_file.c` could be extended to handle the new `X509_L_MEM_LOAD` control command.

Current file-based symbols are:

    # define X509_L_FILE_LOAD        1
    # define X509_LOOKUP_load_file(x,name,type) \
                    X509_LOOKUP_ctrl((x),X509_L_FILE_LOAD,(name),(long)(type),NULL)
    
    X509_LOOKUP_METHOD *X509_LOOKUP_file(void);
    int X509_load_cert_file(X509_LOOKUP *ctx, const char *file, int type);
    int X509_load_crl_file(X509_LOOKUP *ctx, const char *file, int type);
    int X509_load_cert_crl_file(X509_LOOKUP *ctx, const char *file, int type);

New memory-based symbols that I propose:

    // control command for dispatch via X509_LOOKUP_ctrl() or possibly switch in by_file_ctrl()
    # define X509_L_MEM_LOAD        3
    # define X509_LOOKUP_load_mem(x,name,type) \
                    X509_LOOKUP_ctrl((x),X509_L_MEM_LOAD,(name),(long)(type),NULL)
    
    // could return the same structure as X509_LOOKUP_file()
    X509_LOOKUP_METHOD *X509_LOOKUP_mem(void);
    
    int X509_load_cert_mem(X509_LOOKUP *ctx, const char *data, int type);
    int X509_load_crl_mem(X509_LOOKUP *ctx, const char *data, int type);
    int X509_load_cert_crl_mem(X509_LOOKUP *ctx, const char *data, int type);
    
    int X509_load_cert(X509_LOOKUP *ctx, BIO *in, int type);
    int X509_load_crl(X509_LOOKUP *ctx, BIO *in, int type);
    int X509_load_cert_crl(X509_LOOKUP *ctx, BIO *in, int type);

The only caveat I see is that the `const char *data` buffer must be null-terminated.

**Comment by davidben** [comment]
Alternatively, one can simply write a loop to call `PEM_read_bio_X509_AUX` over the input and then `X509_STORE_add_cert` each one. If OpenSSL wants to add a convenience function, I suppose that works, but I would not recommend doing this as an `X509_LOOKUP`. A function that just does that loop is fine.

But this is purely a convenience function. The application can already just write the loop themselves.

**Comment by nhorman** [comment]
yeah, barring a community submitted PR, I don't think we need an extra convenience function here.  Theres several existing solutions.

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
